<h3>Quick basics just to level set:</h3>
Native math in Python

In [38]:
a = 10
b = 2
c = 3

(a*c)/b

15.0

In [39]:
(a*c)**b

900

In [40]:
# Using a simple "f" string (formatted string):

print(f"""

More Arithmetic Natively In Python:

regular division:                   100 /  7 = {100/7}
regular floor divid (integer part): 100 // 7 = {100//7}
modulus (or remainder part):        100 %  7 = {100 % 7}
""")



More Arithmetic Natively In Python:

regular division:                   100 /  7 = 14.285714285714286
regular floor divid (integer part): 100 // 7 = 14
modulus (or remainder part):        100 %  7 = 2



In [41]:
7*14+2

100

<h3>Loops & Lists:</h3>
Python lists (<i>which are not actually arrays</i>) and basic 'loops'.  Note: this is <u>not</u> how a Data Engineer wants to manipulate data.


In [42]:
a = ['a1', 'a2', 'ab', 'a3']
b = ['b1', 'b2', 'ab', 'b3']

for i in a:
    for j in b:
        if i == j:
            print('The common elements are: {}'.format(i))

The common elements are: ab


<h3>Set Theory Math:</h3>
Python has some nice stats functions natively available. <u>But this is an illustration of what a Data Engineer is not going to use all on it's own.</u>

In [43]:
a = ['a1', 'a2', 'ab', 'a3']
b = ['b1', 'b2', 'ab', 'b3']

# create statistical sets from lists a & b
set1 = set(a)
set2 = set(b)

set1.intersection(set2)

{'ab'}

In [44]:
set1 & set2

{'ab'}

In [45]:
# set1.difference(set2)
set1 - set2

{'a1', 'a2', 'a3'}

In [46]:
# set1.uionion(set2)
set1 | set2

{'a1', 'a2', 'a3', 'ab', 'b1', 'b2', 'b3'}

For more advanced functions not natively available in Python, we import well known libraries.  Numpy is pretty much a pre-req in nearly all cases anyway, and used by other libraries.

In [47]:
import numpy as np

a = 10
b = 2
c = 3

np.sqrt((a*c)**b)

30.0

It's sort of like adding a scientific calculator to Python.

In [48]:
print(np.sin(np.radians(45)))
print(np.sin((45/180)*np.pi))

0.7071067811865475
0.7071067811865475


It's a little more powerful than just adding a scientific calculator to Python.  
<p>
Let's take a native python list of numbers, convert it to an numpy array.  Why a numpy array? Because many advanced functions know how to compute element-wise across them (given numpy is also designed for matrix math).
<p>
Notice we are not writting loops! The math library is "<u>vectorized</u>" and runs in high performance compiled code written in C under the covers.

In [71]:
a = [15, 30, 45, 60, 90, -90]
d = np.array(a)

np.sin(np.radians(d))

array([ 0.25881905,  0.5       ,  0.70710678,  0.8660254 ,  1.        ,
       -1.        ])

<h3>DataFrames</h3>
Pandas is the Python DataFrame tool of choice.  Spark as a similar dataframe libary that pretty much works the same. 
<p>
Note we're not opening a CSV file then looping through file handles and writting code, we just call methods within the library.
<p>
<i>It's also written in C and compiled for high performance.</i>

In [50]:
import pandas as pd

df = pd.read_csv('https://raw.githubusercontent.com/plotly/datasets/master/finance-charts-apple.csv')

print(df)

           Date   AAPL.Open   AAPL.High    AAPL.Low  AAPL.Close  AAPL.Volume  \
0    2015-02-17  127.489998  128.880005  126.919998  127.830002     63152400   
1    2015-02-18  127.629997  128.779999  127.449997  128.720001     44891700   
2    2015-02-19  128.479996  129.029999  128.330002  128.449997     37362400   
3    2015-02-20  128.619995  129.500000  128.050003  129.500000     48948400   
4    2015-02-23  130.020004  133.000000  129.660004  133.000000     70974100   
..          ...         ...         ...         ...         ...          ...   
501  2017-02-10  132.460007  132.940002  132.050003  132.119995     20065500   
502  2017-02-13  133.080002  133.820007  132.750000  133.289993     23035400   
503  2017-02-14  133.470001  135.089996  133.250000  135.020004     32815500   
504  2017-02-15  135.520004  136.270004  134.619995  135.509995     35501600   
505  2017-02-16  135.669998  135.899994  134.839996  135.350006     22118000   

     AAPL.Adjusted          dn        m

In [51]:
# display(df)
df

Date   AAPL.Open   AAPL.High    AAPL.Low  AAPL.Close  AAPL.Volume  \
0    2015-02-17  127.489998  128.880005  126.919998  127.830002     63152400   
1    2015-02-18  127.629997  128.779999  127.449997  128.720001     44891700   
2    2015-02-19  128.479996  129.029999  128.330002  128.449997     37362400   
3    2015-02-20  128.619995  129.500000  128.050003  129.500000     48948400   
4    2015-02-23  130.020004  133.000000  129.660004  133.000000     70974100   
..          ...         ...         ...         ...         ...          ...   
501  2017-02-10  132.460007  132.940002  132.050003  132.119995     20065500   
502  2017-02-13  133.080002  133.820007  132.750000  133.289993     23035400   
503  2017-02-14  133.470001  135.089996  133.250000  135.020004     32815500   
504  2017-02-15  135.520004  136.270004  134.619995  135.509995     35501600   
505  2017-02-16  135.669998  135.899994  134.839996  135.350006     22118000   

     AAPL.Adjusted          dn        mavg          up   direction  
0       122.905254  106.741052  117.927667  129.114281  Increasing  
1       123.760965  107.842423  118.940333  130.038244  Increasing  
2       123.501363  108.894245  119.889167  130.884089  Decreasing  
3       124.510914  109.785449  120.763500  131.741551  Increasing  
4       127.876074  110.372516  121.720167  133.067817  Increasing  
..             ...         ...         ...         ...         ...  
501     132.119995  114.494004  124.498666  134.503328  Decreasing  
502     133.289993  114.820798  125.205166  135.589534  Increasing  
503     135.020004  115.175718  125.953499  136.731280  Increasing  
504     135.509995  115.545035  126.723499  137.901963  Decreasing  
505     135.350006  116.203299  127.504333  138.805366  Decreasing  

[506 rows x 11 columns]

In [52]:
df.describe()

AAPL.Open   AAPL.High    AAPL.Low  AAPL.Close   AAPL.Volume  \
count  506.000000  506.000000  506.000000  506.000000  5.060000e+02   
mean   112.935000  113.919447  111.942016  112.958340  4.317842e+07   
std     11.287490   11.251892   11.263687   11.244744  1.985253e+07   
min     90.000000   91.669998   89.470001   90.339996  1.147590e+07   
25%    105.482500  106.349999  104.657501  105.672499  2.974240e+07   
50%    112.889999  114.145001  111.800003  113.025002  3.747460e+07   
75%    122.267498  123.497500  121.599998  122.179998  5.076395e+07   
max    135.669998  136.270004  134.839996  135.509995  1.622063e+08   

       AAPL.Adjusted          dn        mavg          up  
count     506.000000  506.000000  506.000000  506.000000  
mean      110.459312  107.311385  112.739865  118.168345  
std        10.537529   11.095804   10.595315   10.670752  
min        89.008370   85.508858   94.047166   97.572721  
25%       103.484803   97.011245  104.954875  111.052267  
50%       110.821123  107.351628  112.799750  118.472542  
75%       119.255457  114.812152  121.889416  128.515793  
max       135.509995  127.289258  129.845000  138.805366

In [53]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Date           506 non-null    object 
 1   AAPL.Open      506 non-null    float64
 2   AAPL.High      506 non-null    float64
 3   AAPL.Low       506 non-null    float64
 4   AAPL.Close     506 non-null    float64
 5   AAPL.Volume    506 non-null    int64  
 6   AAPL.Adjusted  506 non-null    float64
 7   dn             506 non-null    float64
 8   mavg           506 non-null    float64
 9   up             506 non-null    float64
 10  direction      506 non-null    object 
dtypes: float64(8), int64(1), object(2)
memory usage: 43.6+ KB


And if you want a few columns, I can reference it as follows:

In [54]:
df['AAPL.Close']


0      127.830002
1      128.720001
2      128.449997
3      129.500000
4      133.000000
          ...    
501    132.119995
502    133.289993
503    135.020004
504    135.509995
505    135.350006
Name: AAPL.Close, Length: 506, dtype: float64

In [55]:
# Or a few columns:
df[['Date', 'AAPL.Close', 'AAPL.Volume']]

Date  AAPL.Close  AAPL.Volume
0    2015-02-17  127.830002     63152400
1    2015-02-18  128.720001     44891700
2    2015-02-19  128.449997     37362400
3    2015-02-20  129.500000     48948400
4    2015-02-23  133.000000     70974100
..          ...         ...          ...
501  2017-02-10  132.119995     20065500
502  2017-02-13  133.289993     23035400
503  2017-02-14  135.020004     32815500
504  2017-02-15  135.509995     35501600
505  2017-02-16  135.350006     22118000

[506 rows x 3 columns]

<h2>Visualize</h2>
While poking around the data, I might want some plots.  While pandas has ploting included, and historically many have used 'matplotlib', plotly produces nice interactive charts.  It's well supported with tons of online code samples.

In [56]:
import plotly.graph_objects as go
import plotly.io as pio
from datetime import datetime

fig = go.Figure()
fig.add_trace(
    go.Candlestick(x=df['Date'], 
        open=df['AAPL.Open'], 
        high=df['AAPL.High'], 
        low=df['AAPL.Low'], 
        close=df['AAPL.Close'])
)
fig.show()


<h2>Manipulate</h2>
Let's say, we want to compute a time-series % change in closing price &mdash; without looping through the dataframe and writing the math within the loop<br>
(i.e. it's <i>vectorized</i> just like the Numpy library).

In [57]:
df['AAPL.Close'].pct_change(periods=1)*100

0           NaN
1      0.696236
2     -0.209761
3      0.817441
4      2.702703
         ...   
501   -0.226554
502    0.885557
503    1.297930
504    0.362903
505   -0.118064
Name: AAPL.Close, Length: 506, dtype: float64

<h4>And if you want to added it back to the dataframe... note the last column.</h4>

In [58]:
df['%_chg_Close'] = df['AAPL.Close'].pct_change()*100

df

Date   AAPL.Open   AAPL.High    AAPL.Low  AAPL.Close  AAPL.Volume  \
0    2015-02-17  127.489998  128.880005  126.919998  127.830002     63152400   
1    2015-02-18  127.629997  128.779999  127.449997  128.720001     44891700   
2    2015-02-19  128.479996  129.029999  128.330002  128.449997     37362400   
3    2015-02-20  128.619995  129.500000  128.050003  129.500000     48948400   
4    2015-02-23  130.020004  133.000000  129.660004  133.000000     70974100   
..          ...         ...         ...         ...         ...          ...   
501  2017-02-10  132.460007  132.940002  132.050003  132.119995     20065500   
502  2017-02-13  133.080002  133.820007  132.750000  133.289993     23035400   
503  2017-02-14  133.470001  135.089996  133.250000  135.020004     32815500   
504  2017-02-15  135.520004  136.270004  134.619995  135.509995     35501600   
505  2017-02-16  135.669998  135.899994  134.839996  135.350006     22118000   

     AAPL.Adjusted          dn        mavg          up   direction  \
0       122.905254  106.741052  117.927667  129.114281  Increasing   
1       123.760965  107.842423  118.940333  130.038244  Increasing   
2       123.501363  108.894245  119.889167  130.884089  Decreasing   
3       124.510914  109.785449  120.763500  131.741551  Increasing   
4       127.876074  110.372516  121.720167  133.067817  Increasing   
..             ...         ...         ...         ...         ...   
501     132.119995  114.494004  124.498666  134.503328  Decreasing   
502     133.289993  114.820798  125.205166  135.589534  Increasing   
503     135.020004  115.175718  125.953499  136.731280  Increasing   
504     135.509995  115.545035  126.723499  137.901963  Decreasing   
505     135.350006  116.203299  127.504333  138.805366  Decreasing   

     %_chg_Close  
0            NaN  
1       0.696236  
2      -0.209761  
3       0.817441  
4       2.702703  
..           ...  
501    -0.226554  
502     0.885557  
503     1.297930  
504     0.362903  
505    -0.118064  

[506 rows x 12 columns]

<h4>Let's throw it on the plot on a secondary scale... a simple google search gives samples adapted below.</h4>

In [59]:
import plotly.graph_objects as go
import plotly.io as pio
from datetime import datetime

from plotly.subplots import make_subplots

fig = go.Figure()
fig = make_subplots(specs=[[{"secondary_y": True}]])

fig.add_trace(
    go.Candlestick(x=df['Date'], 
        open=df['AAPL.Open'], 
        high=df['AAPL.High'], 
        low=df['AAPL.Low'], 
        close=df['AAPL.Close']), 
    secondary_y=False
)
fig.add_trace(
    go.Scatter(x=df['Date'], y=df['%_chg_Close'], mode='lines', opacity=0.1, marker_color='blue'), 
    secondary_y=True
)
fig.show()


And if I'd rather take a moving average of the % change column, do the following. Note: we probably would never do this on the % change in real life.

In [60]:
df['%_chg_Close'] = df['AAPL.Close'].pct_change()*100
df['SMA7_%_chg_Close'] = df['%_chg_Close'].rolling(7).mean()

fig = go.Figure()
fig = make_subplots(specs=[[{"secondary_y": True}]])

fig.add_trace(
    go.Candlestick(x=df['Date'], 
        open=df['AAPL.Open'], 
        high=df['AAPL.High'], 
        low=df['AAPL.Low'], 
        close=df['AAPL.Close']), 
    secondary_y=False
)
fig.add_trace(
    go.Scatter(x=df['Date'], y=df['SMA7_%_chg_Close'], mode='lines', opacity=0.1, marker_color='blue'), 
    secondary_y=True
)
fig.show()


In [61]:
df['%_chg_Close'] = df['AAPL.Close'].pct_change(periods=5)*100

fig = go.Figure()
fig = make_subplots(specs=[[{"secondary_y": True}]])

fig.add_trace(
    go.Candlestick(x=df['Date'], 
        open=df['AAPL.Open'], 
        high=df['AAPL.High'], 
        low=df['AAPL.Low'], 
        close=df['AAPL.Close']), 
    secondary_y=False
)
fig.add_trace(
    go.Scatter(x=df['Date'], y=df['%_chg_Close'], mode='lines', opacity=0.1, marker_color='blue'), 
    secondary_y=True
)
fig.show()


<h2>Data Prep</h2>
<h4>
Let's look at some fligth data and assume our job is to clean up some of the data for a Data Science person to run machine learning trials on.
</h4>

In [62]:
df = pd.read_csv(
    'https://raw.githubusercontent.com/yankev/testing/master/datasets/nycflights.csv')

print(f'row count = {len(df)}')
df.sample(10)

row count = 336776


Unnamed: 0  year  month  day  dep_time  dep_delay  arr_time  \
209169      209170  2013      5   17    1851.0       22.0    2207.0   
55412        55413  2013     10   31    1306.0        6.0    1434.0   
88965        88966  2013     12    6       NaN        NaN       NaN   
220825      220826  2013      5   30    1552.0        2.0    1936.0   
115229      115230  2013      2    5    1613.0       23.0    1816.0   
217552      217553  2013      5   27     912.0       -6.0    1145.0   
93893        93894  2013     12   12    1317.0        2.0    1448.0   
326634      326635  2013      9   19    2148.0       -7.0    2256.0   
61715        61716  2013     11    7    1054.0        4.0    1238.0   
185322      185323  2013      4   22    1242.0       51.0    1559.0   

        arr_delay carrier tailnum  flight origin dest  air_time  distance  \
209169       23.0      UA  N38727    1152    EWR  PDX     324.0      2434   
55412         9.0      WN  N788SA    3937    EWR  BNA     117.0       748   
88965         NaN      AA  N4WKAA     300    EWR  DFW       NaN      1372   
220825       29.0      DL  N909DL    1982    LGA  MIA     141.0      1096   
115229       -4.0      9E  N935XJ    3355    JFK  MSP     158.0      1029   
217552      -19.0      B6  N632JB     183    JFK  MCO     127.0       944   
93893        -2.0      AA  N505AA     331    LGA  ORD     119.0       733   
326634      -11.0      UA  N821UA     523    EWR  BOS      48.0       200   
61715       -12.0      MQ  N521MQ    3689    LGA  DTW      80.0       502   
185322      105.0      FL  N926AT     347    LGA  ATL     115.0       762   

        hour  minute  
209169  18.0    51.0  
55412   13.0     6.0  
88965    NaN     NaN  
220825  15.0    52.0  
115229  16.0    13.0  
217552   9.0    12.0  
93893   13.0    17.0  
326634  21.0    48.0  
61715   10.0    54.0  
185322  12.0    42.0

In [63]:
df.drop(df.columns[0], axis=1, inplace=True)

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 336776 entries, 0 to 336775
Data columns (total 16 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   year       336776 non-null  int64  
 1   month      336776 non-null  int64  
 2   day        336776 non-null  int64  
 3   dep_time   328521 non-null  float64
 4   dep_delay  328521 non-null  float64
 5   arr_time   328063 non-null  float64
 6   arr_delay  327346 non-null  float64
 7   carrier    336776 non-null  object 
 8   tailnum    334264 non-null  object 
 9   flight     336776 non-null  int64  
 10  origin     336776 non-null  object 
 11  dest       336776 non-null  object 
 12  air_time   327346 non-null  float64
 13  distance   336776 non-null  int64  
 14  hour       328521 non-null  float64
 15  minute     328521 non-null  float64
dtypes: float64(7), int64(5), object(4)
memory usage: 41.1+ MB


We notice a difference in column counts and realize we have empty cells.  There's more than enough data to go around, so lets take a quick and dirty approach of dropping NaN's and also duplicate rows (just incase).

In [64]:
df.dropna(axis='index', how='any', inplace=True)
df.drop_duplicates(inplace=True)

df

year  month  day  dep_time  dep_delay  arr_time  arr_delay carrier  \
0       2013      1    1     517.0        2.0     830.0       11.0      UA   
1       2013      1    1     533.0        4.0     850.0       20.0      UA   
2       2013      1    1     542.0        2.0     923.0       33.0      AA   
3       2013      1    1     544.0       -1.0    1004.0      -18.0      B6   
4       2013      1    1     554.0       -6.0     812.0      -25.0      DL   
...      ...    ...  ...       ...        ...       ...        ...     ...   
336765  2013      9   30    2240.0       -5.0    2334.0      -17.0      B6   
336766  2013      9   30    2240.0      -10.0    2347.0      -20.0      B6   
336767  2013      9   30    2241.0       -5.0    2345.0      -16.0      B6   
336768  2013      9   30    2307.0       12.0    2359.0        1.0      B6   
336769  2013      9   30    2349.0      -10.0     325.0      -25.0      B6   

       tailnum  flight origin dest  air_time  distance  hour  minute  
0       N14228    1545    EWR  IAH     227.0      1400   5.0    17.0  
1       N24211    1714    LGA  IAH     227.0      1416   5.0    33.0  
2       N619AA    1141    JFK  MIA     160.0      1089   5.0    42.0  
3       N804JB     725    JFK  BQN     183.0      1576   5.0    44.0  
4       N668DN     461    LGA  ATL     116.0       762   5.0    54.0  
...        ...     ...    ...  ...       ...       ...   ...     ...  
336765  N354JB    1816    JFK  SYR      41.0       209  22.0    40.0  
336766  N281JB    2002    JFK  BUF      52.0       301  22.0    40.0  
336767  N346JB     486    JFK  ROC      47.0       264  22.0    41.0  
336768  N565JB     718    JFK  BOS      33.0       187  23.0     7.0  
336769  N516JB     745    JFK  PSE     196.0      1617  23.0    49.0  

[327346 rows x 16 columns]

<h2>Transform</h2>
In Machine Learning, we typically cannot deal with 'text labels'.  Often it is necessary to encode them into numbers.  We can easily grab an array of unique air carriers.

In [65]:
a = df.carrier.unique()

a

array(['UA', 'AA', 'B6', 'DL', 'EV', 'MQ', 'US', 'WN', 'VX', 'FL', 'AS',
       '9E', 'F9', 'HA', 'YV', 'OO'], dtype=object)

We see there are 16 unique carriers in the 327,346 rows and it took less than 1 second

In [66]:
len(a)

16

We'll use the length, and generate a list of numbers to use as our encoding.

In [67]:
b = list((range(101, 101 + len(a), 1)))

b

[101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116]

In [68]:
carrier_codes = dict(zip(a, b))

carrier_codes

{'UA': 101,
 'AA': 102,
 'B6': 103,
 'DL': 104,
 'EV': 105,
 'MQ': 106,
 'US': 107,
 'WN': 108,
 'VX': 109,
 'FL': 110,
 'AS': 111,
 '9E': 112,
 'F9': 113,
 'HA': 114,
 'YV': 115,
 'OO': 116}

In [69]:
df['carrier'] = df.carrier.map(carrier_codes)

df.sample(10)

year  month  day  dep_time  dep_delay  arr_time  arr_delay  carrier  \
286950  2013      8    8    1033.0       -5.0    1146.0      -16.0      105   
177328  2013      4   13    2104.0      -10.0    2253.0      -13.0      105   
133218  2013      2   25    1852.0       -3.0    2208.0       14.0      101   
324680  2013      9   17    2108.0       -5.0    2247.0      -20.0      105   
12168   2013      1   14    2057.0       -3.0    2218.0        6.0      105   
298039  2013      8   20     556.0       -4.0     700.0      -20.0      101   
127116  2013      2   19     705.0        0.0     932.0        4.0      101   
68253   2013     11   14    1109.0       -6.0    1226.0      -26.0      105   
203830  2013      5   12    1022.0       -8.0    1133.0      -37.0      102   
49118   2013     10   24    1614.0       -1.0    1743.0       -4.0      103   

       tailnum  flight origin dest  air_time  distance  hour  minute  
286950  N14991    4510    EWR  MKE     112.0       725  10.0    33.0  
177328  N13988    4700    EWR  CLT      89.0       529  21.0     4.0  
133218  N526UA     258    LGA  IAH     222.0      1416  18.0    52.0  
324680  N11535    4700    EWR  CLT      77.0       529  21.0     8.0  
12168   N19554    4119    EWR  RIC      60.0       277  20.0    57.0  
298039  N81449    1686    EWR  BOS      46.0       200   5.0    56.0  
127116  N427UA     245    EWR  DEN     230.0      1605   7.0     5.0  
68253   N753EV    5273    LGA  PIT      54.0       335  11.0     9.0  
203830  N573AA     321    LGA  ORD     105.0       733  10.0    22.0  
49118   N192JB    2302    JFK  BUF      58.0       301  16.0    14.0